mlflow/llama_index/tracer.py

```diff
@@ -293,3 +340,99 @@ def _extract_token_usage(
+            or isinstance(result, StreamingAgentChatResponse)
+            or (isinstance(result, TaskStepOutput) and self.is_streaming_result(result.output))
+        )
+
+    def register_stream_span(self, span: LiveSpan, result: Any) -> bool:
+        """
+        Register the pending streaming span with the associated generator.
+
+        Args:
+            span: The span that has a streaming output.
+            result: The streaming result that is being processed.
+
+        Returns:
+            True if the span is registered successfully, False otherwise.
+        """
+        if inspect.isgenerator(result):
+            stream = result
+        elif isinstance(result, (StreamingResponse, AsyncStreamingResponse)):
+            stream = result.response_gen
+        elif isinstance(result, StreamingAgentChatResponse):
+            stream = result.chat_stream
+        elif isinstance(result, TaskStepOutput):
+            stream = result.output.chat_stream
+        else:
+            _logger.warning(
+                f"Unsupported streaming response type: {type(result)}. The span will "
+                "be ended immediately without waiting for the stream to finish."
+            )
+            return False
+
+        if inspect.getgeneratorstate(stream) == inspect.GEN_CLOSED:
+            # Not registering the span because the generator is already exhausted
+            return False
+
+        self._span_id_to_span_and_gen[span.span_id] = (span, stream)
+        return True
+
+    def resolve(self, span: LiveSpan, event: _StreamEndEvents):
+        """
+        Finish the streaming span and recursively resolve the parent spans that
+        returns the same (or derived) stream.
+        """
+        _, stream = self._span_id_to_span_and_gen.pop(span.span_id, (None, None))
+        if not stream:
+            return
+
+        if isinstance(event, (LLMChatEndEvent, LLMCompletionEndEvent)):
+            outputs = event.response
+            status = SpanStatusCode.OK
+        elif isinstance(event, ExceptionEvent):
+            outputs = None
+            status = SpanStatusCode.ERROR
+            span.add_event(SpanEvent.from_exception(event.exception))
+        else:
+            raise ValueError(f"Unsupported event type to resolve streaming: {type(event)}")
+
+        _end_span(client=self._client, span=span, status=status, outputs=outputs)
+
+        # Extract the complete text from the event.
+        if isinstance(outputs, ChatResponse):
+            output_text = outputs.message.content
+        elif isinstance(outputs, CompletionResponse):
+            output_text = outputs.response.text
+        else:
+            output_text = None
+
+        # Recursively resolve the parent spans that are also waiting for the same token
+        # stream to be exhausted.
+        while span.parent_id in self._span_id_to_span_and_gen:
+            span, stream = self._span_id_to_span_and_gen.pop(span.parent_id)
+            if stream is not None:
```

Comment:
What does it mean when stream is None but we don't end the span?

Reply:
Great catch! I ported the same logic above but actually L433 is incorrect. The intention is just checking if the span and stream is popped. So we check if the popped item is `None`, but the stream should always be non null.